cmd/operator-opamp-bridge/internal/proxy/server.go

```diff
@@ -0,0 +1,185 @@
+)
+
+type Server interface {
+	GetConfigurations() map[uuid.UUID]*protobufs.EffectiveConfig
+	GetHealth() map[uuid.UUID]*protobufs.ComponentHealth
+	HasUpdates() <-chan struct{}
+}
+
+var _ Server = &OpAMPProxy{}
+
+type OpAMPProxy struct {
+	server.OpAMPServer
+
+	// endpoint for service
+	endpoint string
+
+	// logger
+	logger logr.Logger
+
+	// updatesChan indicates if there any new agents
+	updatesChan chan struct{}
+	// internal state
+	mux        sync.RWMutex
+	agentsById map[uuid.UUID]*Agent
+	// connections map is required because that's the only way we know to remove an agent.
+	connections map[types.Connection]map[uuid.UUID]bool
+}
+
+func NewOpAMPProxy(l logr.Logger, endpoint string) *OpAMPProxy {
+	return &OpAMPProxy{
+		logger:      l,
+		OpAMPServer: server.New(logger.NewLogger(l)),
+		endpoint:    endpoint,
+		agentsById:  map[uuid.UUID]*Agent{},
+		connections: map[types.Connection]map[uuid.UUID]bool{},
+		updatesChan: make(chan struct{}, 1),
+	}
+}
+
+func (s *OpAMPProxy) Start() error {
+	settings := server.StartSettings{
+		Settings: server.Settings{
+			CustomCapabilities: []string{},
+			Callbacks: server.CallbacksStruct{
+				OnConnectingFunc: func(request *http.Request) types.ConnectionResponse {
+					return types.ConnectionResponse{
+						Accept: true,
+						ConnectionCallbacks: server.ConnectionCallbacksStruct{
+							OnMessageFunc:         s.onMessage,
+							OnConnectionCloseFunc: s.onDisconnect,
+						},
+					}
+				},
+			},
+		},
+		ListenEndpoint: s.endpoint,
+		HTTPMiddleware: otelhttp.NewMiddleware("/v1/opamp"),
+	}
+	// TODO: In the future we will probably want some TLS configuration.
+	// tlsConfig, err := internal.CreateServerTLSConfig(
+	// 	"../../certs/certs/ca.cert.pem",
+	// 	"../../certs/server_certs/server.cert.pem",
+	// 	"../../certs/server_certs/server.key.pem",
+	// )
+	// if err != nil {
+	// 	srv.logger.Debugf(context.Background(), "Could not load TLS config, working without TLS: %v", err.Error())
+	// }
+	// settings.TLSConfig = tlsConfig
+	s.logger.Info("starting opamp server", "address", s.endpoint)
+	return s.OpAMPServer.Start(settings)
+}
+
+func (s *OpAMPProxy) Stop(ctx context.Context) error {
+	return s.OpAMPServer.Stop(ctx)
+}
+
+func (s *OpAMPProxy) onDisconnect(conn types.Connection) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	for instanceId := range s.connections[conn] {
+		delete(s.agentsById, instanceId)
+	}
+	delete(s.connections, conn)
+}
+
+func (s *OpAMPProxy) onMessage(ctx context.Context, conn types.Connection, msg *protobufs.AgentToServer) *protobufs.ServerToAgent {
+	// Start building the response.
+	response := &protobufs.ServerToAgent{}
+
+	instanceId, err := getInstanceId(msg)
+	if err != nil {
+		s.logger.Error(err, "failed to get instance id")
+		response.ErrorResponse = &protobufs.ServerErrorResponse{
+			ErrorMessage: err.Error(),
+		}
+		return response
+	}
+	s.logger.V(5).Info("received message", "instance ID", instanceId)
+
+	agentUpdated := false
```

Comment:
This should be called `agentAdded` if that's the only condition it actually tracks.

Reply:
this actually does track updates from the agent itself. the UpdateStatus checks if things like the description changed.